xds/internal/xdsclient/transport/transport.go

```diff
@@ -621,3 +637,68 @@ func (t *Transport) Close() {
 func (t *Transport) ChannelConnectivityStateForTesting() connectivity.State {
 	return t.cc.GetState()
 }
+
+// ADSFlowControl implements ADS stream level flow control that enables the
+// transport to block the reading of the next message off of the stream until
+// the previous update is consumed by all watchers.
+//
+// The lifetime of the flow control is tied to the lifetime of the stream.
+//
+// New instances must be created with a call to NewADSStreamFlowControl.
+type ADSFlowControl struct {
+	logger *grpclog.PrefixLogger
+
+	// Count of watchers yet to consume the most recent update.
+	pending atomic.Int64
+	// Channel used to notify when all the watchers have consumed the most
+	// recent update. Wait() blocks on reading a value from this channel.
+	readyCh chan struct{}
+}
+
+// NewADSStreamFlowControl returns a new ADSFlowControl.
+func NewADSStreamFlowControl() *ADSFlowControl {
+	return &ADSFlowControl{readyCh: make(chan struct{}, 1)}
+}
+
+// Add increments the number of watchers (by one) who are yet to consume the
+// most recent update received on the ADS stream.
+func (fc *ADSFlowControl) Add() {
+	fc.pending.Add(1)
+}
+
+// Wait blocks until all the watchers have consumed the most recent update and
+// returns true. If the context expires before that, it returns false.
+func (fc *ADSFlowControl) Wait(ctx context.Context) bool {
+	// If there are no watchers with pending updates, there is no need to block.
+	if n := fc.pending.Load(); n == 0 {
+		// If all watchers
+		select {
+		case <-fc.readyCh:
+		default:
+		}
```

Comment:
Why does this need to even been here? If this is zero when the program flow calls wait, nothing will ever send on the readyCh since nothing will ever invoke onDone. Also why half of a comment above?

Reply:
Completed the comment. Hope this makes it clear.

Reply:
Ah that makes sense. It could finish before it gets here.